File: 0039_combination-sum/solution.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        end = len(candidates)

        def dfs(start, path, target):
            if target < 0:  # 不符合条件的剪枝
                return
            if target == 0:  # 将符合条件的保存
                res.append(path)
                return
            for i in range(start, end):
                cur = candidates[i]
                dfs(i, path + [cur], target - cur)

        dfs(0, [], target)
        return res

File: 0039_combination-sum/solution.go
package main

func combinationSum(candidates []int, target int) [][]int {
	res := [][]int{}
	end := len(candidates)
	if end == 0 {
		return res
	}
	var dfs func(start int, temp []int, tmpTarget int)
	dfs = func(start int, temp []int, tmpTarget int) {
		if tmpTarget < 0 {
			return
		}
		if tmpTarget == 0 {
			newTmp := make([]int, len(temp))
			copy(newTmp, temp)
			res = append(res, newTmp)
			return
		}
		for i := start; i < end; i++ {
			temp = append(temp, candidates[i])
			dfs(i, temp, tmpTarget-candidates[i])
			temp = temp[:len(temp)-1]
		}

	}
	dfs(0, []int{}, target)
	return res

}
